File: backend/services/workspace/src/main/kotlin/dev/kuku/taskinator/domains/project/internal/ProjectServiceImpl.kt
package dev.kuku.taskinator.domains.project.internal

import com.github.f4b6a3.uuid.UuidCreator
import dev.kuku.taskinator.domains.project.*
import io.github.oshai.kotlinlogging.KotlinLogging
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.UUID

private val log = KotlinLogging.logger {}

/**
 * PROJECT DOMAIN SERVICE: Business Logic & Guardrails
 */
@Service
@Transactional
class ProjectServiceImpl(
    private val projectRepo: ProjectQueries
) : ProjectService {

    private val MAX_LIMIT = 100

    override fun createProject(
        userId: String,
        projectName: String,
        projectDescription: String?,
        projectId: String?
    ): ProjectInfo? {
        log.info { "Creating Project ${projectName} for user $userId" }
        try {
            val finalId = projectId ?: UuidCreator.getTimeOrderedWithRandom().toString()
            return projectRepo.insertProject(
                finalId,
                projectName,
                userId,
                projectDescription,
                UUID.randomUUID().toString()
            )
        } catch (e: ProjectNameConflictException) {
            log.warn { "Creation failed: duplicate name" }
            throw e
        }
    }

    override fun renameProject(userId: String, projectId: String, toUpdate: ProjectFieldsToUpdate) {
        log.info { "Renaming project $projectId" }
        try {
            projectRepo.updateProject(projectId, userId, toUpdate)
        } catch (e: Exception) {
            when (e) {
                is ProjectConcurrencyException,
                is ProjectNameConflictException -> throw e
                else -> throw e
            }
        }
    }

    override fun addProjectMembers(projectId: String, userId: String, memberIds: List<String>) {
        try {
            projectRepo.insertProjectMembers(projectId, userId, memberIds)
        } catch (e: Exception) {
            log.error(e) { "Failed to add project members" }
            throw e
        }
    }

    override fun removeProjectMembers(projectId: String, userId: String, memberIds: List<String>) {
        try {
            projectRepo.deleteProjectMembers(projectId, userId, memberIds)
        } catch (e: Exception) {
            log.error(e) { "Failed to remove project members" }
            throw e
        }
    }

    override fun deleteProject(projectId: String, userId: String, version: Long): Boolean {
        // We do NOT perform a hard delete of the project metadata here.
        // Instead, we verify ownership/version and trigger the ASYNC cleanup.
        // In a real high-scale system, you might mark it as status = 'DELETED' first.
        
        // For now, we perform the metadata delete and member delete synchronously 
        // as they are small tables, but the HEAVY data (Teams/Tasks) will be 
        // handled via the event fired from the controller.
        val deletedRows = projectRepo.deleteProject(projectId, userId, version)

        if (deletedRows == 0) {
            throw ProjectConcurrencyException("Delete failed: version mismatch or unauthorized.")
        }

        log.info { "ProjectService: Metadata deleted for project $projectId" }
        return true
    }

    override fun getProjectById(projectId: String, userId: String): ProjectInfo? {
        return projectRepo.findProjectById(projectId, userId)
    }

    override fun getProjectsByUser(userId: String, limit: Int, offset: Int): List<ProjectInfo> {
        val enforcedLimit = limit.coerceAtMost(MAX_LIMIT)
        return projectRepo.findProjectsByOwner(userId, enforcedLimit, offset)
    }

    override fun getProjectMembers(
        projectId: String,
        userId: String,
        sortBy: ProjectMemberSortKey,
        offset: Int,
        limit: Int
    ): List<ProjectMember> {
        val enforcedLimit = limit.coerceAtMost(MAX_LIMIT)
        return projectRepo.findProjectMembers(projectId, userId, sortBy, offset, enforcedLimit)
    }

    override fun getProjectUserIsPartOf(userId: String, limit: Int, offset: Int): List<String> {
        val enforcedLimit = limit.coerceAtMost(MAX_LIMIT)
        return projectRepo.findProjectIdsByUserMembership(userId, enforcedLimit, offset)
    }
}

File: modular-monolith/src/services/project/internal/ProjectServiceImpl.ts
import type {
    AddProjectMembersParam,
    CreateProjectParam,
    DeleteProjectMembersParam,
    DeleteProjectsParam,
    Project,
    ProjectMember,
    ProjectService,
} from '../ProjectService.ts';
import {
    deleteProjectMembers,
    deleteProjects,
    getProject,
    getProjects,
    insertProject,
    insertProjectMembers,
    insertProjects,
} from './ProjectQueries.ts';

import {
    KAFKA_TOPICS,
    KAFKA_EVENTS,
    createEvent,
} from '../../../utils/kafka.ts';
import { eventBus } from '../../../utils/EventBus.ts';

export class ProjectServiceImpl implements ProjectService {
    async getProjects(userId: string): Promise<Project[]> {
        console.log(`[Project Service] Getting projects for userId: ${userId}`);
        return getProjects(userId);
    }

    async getProject(
        userId: string,
        projectId: string,
    ): Promise<Project | null> {
        return getProject(userId, projectId);
    }

    async destroy(): Promise<void> {
        console.log(`Disconnecting event bus ${this.constructor.name}`);
        await eventBus.destroy();
    }

    async init() {
        console.log(`Initializing event bus ${this.constructor.name}`);
        await eventBus.init();
    }

    async deleteProjectMembers(data: DeleteProjectMembersParam): Promise<void> {
        const deleted = await deleteProjectMembers(data);

        if (!deleted.length) {
            throw new Error('Failed to delete any project members');
        }

        const events = deleted.map((v) =>
            createEvent(KAFKA_EVENTS.PROJECT_MEMBER.DELETED, data.projectId, {
                userId: v.userId,
                projectId: data.projectId,
                actorId: data.userId,
                memberId: v.id,
            }),
        );

        await eventBus.emit(KAFKA_TOPICS.PROJECT_MEMBER, events);
    }

    async deleteProjects(data: DeleteProjectsParam): Promise<void> {
        const deleted = await deleteProjects(data);

        if (!deleted.length) {
            throw new Error('Failed to delete any project');
        }

        const events = deleted.map((project) =>
            createEvent(KAFKA_EVENTS.PROJECT.DELETED, project.id, {
                userId: data.userId,
                projectId: project.id,
            }),
        );

        await eventBus.emit(KAFKA_TOPICS.PROJECT, events);
    }

    async addProjectMembers(
        data: AddProjectMembersParam,
    ): Promise<ProjectMember[]> {
        const added = await insertProjectMembers(data);

        if (!added.length) {
            throw new Error('Failed to add any project members');
        }

        const events = added.map((member) =>
            createEvent(KAFKA_EVENTS.PROJECT_MEMBER.ADDED, data.projectId, {
                projectId: data.projectId,
                userId: member.userId,
                actorId: data.userId,
                memberId: member.id,
            }),
        );

        await eventBus.emit(KAFKA_TOPICS.PROJECT_MEMBER, events);

        return added;
    }

    async createProject(data: CreateProjectParam): Promise<Project | null> {
        const project = await insertProject(data);

        if (!project) {
            throw new Error('Failed to create project');
        }

        await eventBus.emit(
            KAFKA_TOPICS.PROJECT,
            createEvent(KAFKA_EVENTS.PROJECT.CREATED, project.id, {
                projectId: project.id,
                userId: data.userId,
                name: project.name,
            }),
        );

        return project;
    }

    async createProjects(data: CreateProjectParam[]): Promise<Project[]> {
        const projects = await insertProjects(data);

        if (!projects.length) {
            throw new Error('Failed to create projects');
        }

        const events = projects.map((project) =>
            createEvent(KAFKA_EVENTS.PROJECT.CREATED, project.id, {
                projectId: project.id,
                userId: project.userId,
                name: project.name,
            }),
        );

        await eventBus.emit(KAFKA_TOPICS.PROJECT, events);

        return projects;
    }
}
